Cache the list of workers in memory

aoldershaw opened the pull request:
## Changes proposed by this PR

On hush-house and CI, the following queries take up a substantial amount of DB CPU time:

* `WorkerFactory.Workers`
* `WorkerFactory.BuildContainersCountPerWorker`

They are each called every time a container is scheduled - on hush-house, this amounts to ~250,000 times an hour. This worker data can easily fit into memory to significantly reduce the number of round-trips to the DB.

* [x] Extend notification bus to support queueing up multiple notifications and sending payloads
* [x] Add postgres triggers to the `workers` and `containers` tables - these triggers send a notification along the notification bus containing the type of operation (`UPDATE,` `INSERT`, or `DELETE`) and the affected worker name
* [x] Implement a `WorkerCache` that listens for these trigger events and modifies in-memory data structures
    * `WorkerCache` fetches this data every minute as well, to ensure the data doesn't get out of sync

## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->

## Release Note

* Scheduling containers should be more performant by reducing the number of required database calls

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

aoldershaw committed:
B: support queueing notification bus

...and support receiving string payloads from the bus

The current behaviour of the notification bus when a new notification
comes in is:

Try to send that notification to each Go chan:
* If that chan is empty, send along the chan
* If that chan already has a notification waiting to be received, do
  nothing

This is nice because all we previously cared about is that the consumer
is told that something changed. However, if we want to start receiving
data on the notification bus (e.g. triggers that include the ID of the
element that was changed), we'll need to queue several events in case
the consumer isn't fast enough to drain them.

Note that I tagged this commit as behaviour (since it introduces new
behaviour that can be used), but it doesn't yet change anything since
nothing is making use of it - everything is still using the "don't
queue notifications" behaviour, and nothing sends any payloads.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: cache workers in memory

the motivation for this change is to optimize scheduling, since the
WorkerFactory methods (esp. Workers and BuildContainersCountPerWorker)
get called every time a container is scheduled. on hush-house, these two
queries consume a ton of CPU time.

this commit adds a new WorkerCache component that tracks the current
workers (and their corresponding container counts) in memory.
WorkerFactory shells out to WorkerCache when it lists or looks up
workers [1], and when it lists the build containers count.

WorkerCache keeps its in-memory data relatively fresh by listening to
changes in these tables via postgres TRIGGERs. when a worker is
INSERTed, UPDATEd, or DELETEd, the WorkerCache will update its internal
state accordingly. for UPDATE/INSERT, this requires making a DB query
for the single worker that changed.

similarly, we track INSERTs/DELETEs for containers. here, however, we
don't need to do another query - we just increment/decrement a counter.

it's possible that the data can still get out of sync, since it's
possible to miss updates on the notification bus. for this, we
periodically refresh the state once every minute so we don't stay out of
sync for too long.

[1] the exception here is FindWorkersForContainerByOwner, which still
makes a query. we *could* store the list of containers in memory as
well, but it doesn't seem to be a bottleneck in the DB so it didn't seem
worth it.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

xtremerui commented on atc/db/migration/migrations/1625844436_add_worker_cache_triggers.down.sql:
```diff
@@ -0,0 +1,4 @@
+DROP TRIGGER workers_upsert_or_delete_trigger ON workers;
```
could we add a `if exist` here?

xtremerui reviewed (commented):
is the extended notification bus queue for only notifications from that trigger or is it queueing for all? It seems like it is for all notifications. In this case, is there performance trade off (or gain) because of the queue?

xtremerui commented on atc/db/worker_cache.go:
```diff
@@ -0,0 +1,278 @@
+
+	cache.workers[name] = worker
+	if _, ok := cache.workerContainerCounts[name]; !ok {
+		cache.workerContainerCounts[name] = 0
```
when cache is un-sync with DB like this, should it be removed like line 259 or set to 0?

aoldershaw replied:
This condition isn't necessarily about being out-of-sync - if a new worker is inserted, it won't yet have an entry in the `workerContainerCounts` table. Since it can't have containers yet, we add it to the table with 0 containers

aoldershaw commented:
> is the extended notification bus queue for only notifications from that trigger or is it queueing for all? It seems like it is for all notifications. In this case, is there performance trade off (or gain) because of the queue?

The queueing in the notification bus is really just a generalization of the existing behaviour. Previously, all uses of the notification bus were "tell me *if* something has changed". Since we ignore messages that come in when the queue size is exceeded, we achieved this desired behaviour using a queue size of 1 for the listening channel. All that to say, this doesn't impact the behaviour/performance of the existing notification bus users - it just makes setting the queue size explicit.

Now that we're sending messages that are more meaningful than "has something changed", and all of those messages need to be processed individually, we need to allow sending more messages along a single channel, since the consumer of the channel may be slower than the producer (depending on how Go schedules goroutines). This shouldn't be significantly different in terms of performance than with a queue size of 1

aoldershaw committed:
add IF EXISTS to migrations

shouldn't be necessary unless you manually mess with the DB schema, but
doesn't hurt to have

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
